File: packages/js/core/src/interfaces/uri-resolver.ts
import { Uri, Invoker, WrapError } from "../";

import { Tracer } from "@polywrap/tracing-js";
import { Result } from "@polywrap/result";

export interface MaybeUriOrManifest {
  uri?: string | null;
  manifest?: Uint8Array | null;
}

export const module = {
  tryResolveUri: Tracer.traceFunc(
    "core: uri-resolver: tryResolveUri",
    async (
      invoker: Invoker,
      wrapper: Uri,
      uri: Uri
    ): Promise<Result<MaybeUriOrManifest, WrapError>> => {
      return invoker.invoke<MaybeUriOrManifest>({
        uri: wrapper.uri,
        method: `tryResolveUri`,
        args: {
          authority: uri.authority,
          path: uri.path,
        },
      });
    }
  ),
  getFile: Tracer.traceFunc(
    "core: uri-resolver: getFile",
    async (
      invoker: Invoker,
      wrapper: Uri,
      path: string
    ): Promise<Result<Uint8Array | null, WrapError>> => {
      return invoker.invoke<Uint8Array | null>({
        uri: wrapper.uri,
        method: "getFile",
        args: {
          path,
        },
      });
    }
  ),
};

File: packages/py/core/src/interfaces/uri_resolver.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Awaitable

from .. import InvokeApiOptions, InvokeApiResult


@dataclass
class MaybeUriOrManifest:
    uri: str = ""
    manifest: str = ""


class UriResolverInterface:
    @classmethod
    async def try_resolve_uri(cls, invoke: InvokeHandler, api: Uri, uri: Uri) -> Awaitable[MaybeUriOrManifest]:
        return await invoke(
            InvokeApiOptions(
                uri=api.uri,
                method="try_resolve_uri",
                input={
                    "authority": uri.authority,
                    "path": uri.path,
                },
            )
        )

    @classmethod
    async def get_file(cls, invoke: InvokeHandler, api: Uri, path: str) -> Awaitable[InvokeApiResult]:
        return await invoke(
            InvokeApiOptions(
                uri=api.uri,
                method="get_file",
                input={
                    "path": path,
                },
            )
        )
